src/Xamarin.Android.Build.Tasks/Utilities/ApplicationConfigNativeAssemblyGenerator.cs

```diff
@@ -39,287 +113,246 @@ class ApplicationConfigNativeAssemblyGenerator : NativeAssemblyGenerator
 				for (int i = 0; i < NumberOfAssemblyStoresInApks; i++) {
-					size += WriteStructure (output, packed: false, structureWriter: () => WriteAssemblyStore (output));
+					NativeAssemblyGenerator.StructureWriteContext assemblyStoreStruct = generator.AddStructureArrayElement (assemblyStoreArray);
+					generator.WriteStructure (assemblyStoreStruct, emptyStoreData);
 				}
+				generator.WriteSymbol (assemblyStoreArray, label, local: false);
+			} else {
+				generator.WriteEmptySymbol (SymbolType.Object, label, local: false);
 			}
-			WriteStructureSize (output, label, size);
-		}
-
-		uint WriteAssemblyStoreAssembly (StreamWriter output)
-		{
-			// Order of fields and their type must correspond *exactly* to that in
-			// src/monodroid/jni/xamarin-app.hh AssemblyStoreSingleAssemblyRuntimeData structure
-			WriteCommentLine (output, "image_data");
-			uint size = WritePointer (output);
-
-			WriteCommentLine (output, "debug_info_data");
-			size += WritePointer (output);
-
-			WriteCommentLine (output, "config_data");
-			size += WritePointer (output);
-
-			WriteCommentLine (output, "descriptor");
-			size += WritePointer (output);
-
-			output.WriteLine ();
-
-			return size;
 		}
 
-		uint WriteAssemblyStore (StreamWriter output)
+		void WriteBundledAssemblies (NativeAssemblyGenerator generator)
 		{
-			// Order of fields and their type must correspond *exactly* to that in
-			// src/monodroid/jni/xamarin-app.hh AssemblyStoreRuntimeData structure
-			WriteCommentLine (output, "data_start");
-			uint size = WritePointer (output);
-
-			WriteCommentLine (output, "assembly_count");
-			size += WriteData (output, (uint)0);
-
-			WriteCommentLine (output, "assemblies");
-			size += WritePointer (output);
-
-			output.WriteLine ();
+			generator.WriteCommentLine ($"Bundled assembly name buffers, all {BundledAssemblyNameWidth} bytes long");
+			generator.WriteSection (".bss.bundled_assembly_names", SectionFlags.Writable | SectionFlags.Allocatable, SectionType.NoData);
 
-			return size;
-		}
-
-		void WriteBundledAssemblies (StreamWriter output)
-		{
-			output.WriteLine ();
-
-			WriteCommentLine (output, $"Bundled assembly name buffers, all {BundledAssemblyNameWidth} bytes long");
-			WriteSection (output, ".bss.bundled_assembly_names", hasStrings: false, writable: true, nobits: true);
-
-			List<string> name_labels = null;
+			List<NativeAssemblyGenerator.LabeledSymbol>? name_labels = null;
 			if (!HaveAssemblyStore) {
-				name_labels = new List<string> ();
+				name_labels = new List<NativeAssemblyGenerator.LabeledSymbol> ();
 				for (int i = 0; i < NumberOfAssembliesInApk; i++) {
-					string bufferLabel = GetBufferLabel ();
-					WriteBufferAllocation (output, bufferLabel, (uint)BundledAssemblyNameWidth);
-					name_labels.Add (bufferLabel);
+					name_labels.Add (generator.WriteEmptyBuffer ((uint)BundledAssemblyNameWidth, "env.buf"));
```

Comment:
Another one for the naming $DEITYs, "write empty buffer"?  I'm not at all sure what the semantics are based on just looking at this line.

I wonder if `BeginBuffer(…)` would be better?  Unless this is asserting that the buffer is in fact empty, but why would we ever have known empty buffers?  (I have not yet read what `WriteEmptyBuffer()` *does*.  I'm not sure I'll understand what it does when I get there.)

Reply:
I'm really not sure what you're asking for here. If one doesn't know what the buffer does in this context, then I don't want them to touch the code. This is the lowest level generator for a specific set of specific structures that describe the entirety of application config, and modifying the code **requires** understanding what the structures are on the native side. Once that is understood, the terms used will make sense. The buffer is in fact empty, because it just allocates space in the output DSO to be filled at the run time. In this case, the code is used when assembly stores are off and we read the bundled assemblies from the apk, filling in this structure:

```c++

struct XamarinAndroidBundledAssembly final
{
        int32_t  apk_fd;
        uint32_t data_offset;
        uint32_t data_size;
        uint8_t *data;
        uint32_t name_length;
        char    *name;
};
```

the buffer is for the assembly name (as suggested by `BundledAssemblyNameWidth` in the method invocation) and has to be there because `name` has to point to something, we don't allocate data directly. All that information is there, in source code, and I do expect anyone modifying this code to familiarize themselves with the native side...

I'm **really** not sure how do you want to explain all that in a method name?